pkg/apis/servicecatalog/v1alpha1/types.go

```diff
@@ -409,22 +409,31 @@ type ServiceInstance struct {
 	Status ServiceInstanceStatus `json:"status"`
 }
 
-// ServiceInstanceSpec represents the desired state of an Instance.
-type ServiceInstanceSpec struct {
+// DesiredPlan defines the user specification for the desired
+// ServicePlan and ServiceClass. Because there are multiple ways to
+// specify the desired Class/Plan, this structure specifies the
+// allowed ways to specify the intent.
+type DesiredPlan struct {
 	// ExternalClusterServiceClassName is the human-readable name of the
 	// service as reported by the broker. Note that if the broker changes
 	// the name of the ClusterServiceClass, it will not be reflected here,
 	// and to see the current name of the ClusterServiceClass, you should
 	// follow the ClusterServiceClassRef below.
 	//
 	// Immutable.
-	ExternalClusterServiceClassName string `json:"externalClusterServiceClassName"`
+	ExternalClusterServiceClassName string `json:"externalClusterServiceClassName,omitempty"`
 	// ExternalClusterServicePlanName is the human-readable name of the plan
 	// as reported by the broker. Note that if the broker changes the name
 	// of the ClusterServicePlan, it will not be reflected here, and to see
 	// the current name of the ClusterServicePlan, you should follow the
 	// ClusterServicePlanRef below.
 	ExternalClusterServicePlanName string `json:"externalClusterServicePlanName,omitempty"`
+}
+
+// ServiceInstanceSpec represents the desired state of an Instance.
+type ServiceInstanceSpec struct {
+	// Specification of what ServiceClass/ServicePlan is being provisioned.
+	DesiredPlan `json:",inline"`
```

Comment:
Not sure if embedding is better than having a separate `plan` section in JSON/YAML, given that the structure will be growing in the future.

Reply:
My understanding was that for the user if they have to add another layer of indirection, this would be  undesired. I frankly don't feel very strongly either way about this.
Just to be clear though, we can embed the struct and either expose it as inline at the json/yaml layer, or we can make it a field of type and then of course we couldn't 'inline' it at the json/yaml layer, which do you oppose to?

Reply:
I was opposing to both actually, but primarily to `inline` (serialization)

Reply:
Ok, so what would you like to see then? More fields at the ServiceInstanceSpec level both at the yaml and our types?

Reply:
Yep, I would prefer to see:
```
kind: ServiceInstance
spec:
  plan:
    clusterServiceClassName: foo
    cluserServicePlanName: bar
```
and
```
type ServiceInstanceSpec struct {
	Plan PlanReference `json:"plan,omitempty"`
```
I would ask @pmorie on which option does he prefer, and make decision based on that.

Reply:
the extra "plan" wrapper serves no purpose - the entire Spec is about "what plan to use", why add more wrappers?

Reply:
> the entire Spec is about "what plan to use"

Not true. The Instance spec has also parameters and server-generated data (user info, externalId etc). The reason why I want it to be a separate section is because the plan reference is composite (class + plan).

@vaikas-google  I would actually put the other two fields inside the `PlanReference` structure as well:
```go
 	// ClusterServiceClassRef is a reference to the ClusterServiceClass
  	// that the user selected.
 	// This is set by the controller based on ExternalClusterServiceClassName
 	ClusterServiceClassRef *v1.ObjectReference
 	// ClusterServicePlanRef is a reference to the ClusterServicePlan
 	// that the user selected.
 	// This is set by the controller based on ExternalClusterServicePlanName
 	ClusterServicePlanRef *v1.ObjectReference
```
and probably let the user manually set them instead of `ExternalName`s in the future.
Then we can think of `PlanRefence` Type as a composite union type.

> why add more wrappers

What's wrong with extra nesting levels? They are free. Have you seen the Pod spec structure? :)

Reply:
I actually don't insist on adding extra nesting level in JSON, just wanted to clarify my POV.
I do think that the 2 other fields (`ClusterServiceClassRef` and `ClusterServicePlanRef`) should belong to `PlanReference` though, if there are not objections.

Reply:
Yes there are parameters but they're there to help augment information about what plan to use and how to modify its config.  Wrappers make sense to me when there are separations of concerns - e.g. if ServiceInstance was used to store other info not related to the Service & Plan, but it doesn't. Adding the nesting doesn't provide any value to me and just causes more unnecessary typing.  Yes I've seen the Pod spec and there are many different "thoughts" at play in there at once, so grouping things makes sense.  A grouping for a pointer doesn't to me.

Reply:
@duglin last argument towards having an extra section: see `LocalObjectReference`:
```go
// LocalObjectReference contains enough information to let you locate the
// referenced object inside the same namespace.
type LocalObjectReference struct {
	// Name of the referent.
	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
	// TODO: Add other useful fields. apiVersion, kind, uid?
	// +optional
	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
}
```
It could have been easily inlined without any confusion, but for some reason it has a nested `name` field. Yes it has a TODO, which was never addressed :)

Reply:
Yep and that makes sense because its generic - meaning within one Resource you may have more than one "ref" and so you need to know which "Name" goes with which "ref".  I don't think anyone has suggested allowing a single ServiceInstance to point to more than one service/plan at a time.

Our current example:
```
spec:
  externalServiceClassName: user-provided-service
  externalServicePlanName: default
  parameters:
    credentials:
      param-1: value-1
      param-2: value-2
```
Would become:
```
spec:
  DesiredPlan:
    externalServiceClassName: user-provided-service
    externalServicePlanName: default
  parameters:
    credentials:
      param-1: value-1
      param-2: value-2
```
What's the value to the end-user? Was the first one confusing? I don't think so. If anything, I might ask why params isn't part of the "DesiredPlan" since params go hand-in-hand with the plan being specified.

Reply:
Consensus: leave the API as it is, without extra grouping/nesting.